### PR TITLE
Generalizable Text Transformer Usage: Adding HuggingFace Model

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

File: src/open_clip/hf_model.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        super().__init__()
+        if transformers is None:
+            raise RuntimeError("Please `pip install transformers` to use pre-trained HuggingFace models")
+        self.config = AutoConfig.from_pretrained(model_name_or_path)
+        self.transformer = AutoModel.from_pretrained(model_name_or_path)
+        
+        self.pooler = get_pooler(pooler_type)
+        d_model = self.config.hidden_size # TODO: get d_model from config
+        # different models can have different names for it
+        # ?? do we use separate classes for different archs or handle it with helper funcs
```
[the remark this answers is not in the document]
Comment (arampacha) [reply src/open_clip/hf_model.py:63]: Yeah, the question is how to keep this concise. We can organize it as dict of dicts: `arch_name` -> `arch_meta`, where `arch_meta` will hold all relevant info, like default `pooler`, config name mappings etc.

File: src/open_clip/hf_model.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        if (d_model == embedding_dim) and (proj is None): # do we always need a proj?
+            self.proj = nn.Identity()
+        elif proj == 'linear':
+            self.proj == nn.Linear(d_model, embedding_dim, bias=False)
+        elif proj == 'mlp':
+            # TODO: add me
+            pass
+
+    def forward(self, x:TensorType) -> TensorType:
+        # TODO: add attention mask?
```
[the remark this answers is not in the document]
Comment (arampacha) [reply src/open_clip/hf_model.py:73]: yep, I wanted to check if HF models add attention mask for padding tokens by default in no `attention_mask` is provided.
I think causal_mask is added automatically for autoregressive models by HF too

[the remark this answers is not in the document]
Comment (iejMac) [reply src/open_clip/hf_model.py:73]: Do we need any other type of attention masking than casual? If not then we can just always input this attention mask to be safe.

[the remark this answers is not in the document]
Comment (arampacha) [reply src/open_clip/hf_model.py:73]: yeah, we'll need `attention_mask` for pooling anyway
MB we can use the atention_mask returned by tokenizer and avoid recreating it here, if we can make it compatible with the training code
For now lets `build_attention_mask` here using `pad_token_id` for the model

**Comment by rom1504** [comment]
Hey, I'm interested by using this 

What is the current state of this work?
What is missing, what do you expect need to be done next ?

**Comment by iejMac** [comment]
Hi,

Current state - we've implemented a PreTrainedTextEncoder class that has an interface compatible with the TextTransformer class from the grad caching PR. This is roughly equivalent to the first 3 points we listed here - https://github.com/mlfoundations/open_clip/issues/53

What is missing - we now need to integrate this PreTrainedTextEncoder to the CLIPModel and the HF tokenizer into the text processing pipeline.

**Comment by rom1504** [comment]
Can you add something in the readme pointing how to use this ?

**Comment by iejMac** [comment]
@rom1504 currently the working version is in [this branch](https://github.com/iejMac/open_clip/tree/dev-integration) where I added the README section.

FYI: @arampacha detailed why we branched off [here](https://github.com/mlfoundations/open_clip/pull/93#issuecomment-1142593779). I'm not sure we can merge until Ross finishes up his gradient caching PR (which dev-integration has already merged in).

**Comment by rom1504** [comment]
@iejMac if you have some time, I think it would be great to open a PR so we can merge that ?

@rwightman do you think your PR is blocking? do you plan to finish it ?

**Comment by rwightman** [comment]
@iejMac @rom1504 what is the status? is the generalizable text transformer work in a good enough state to merge?

The grad caching turned out to be quite a bit more involved to determine if it is actually equivalent wrt to training results at a half decent scale.... I don't have extra bandwidth (or compute to dedicate) to sort that out.

It's easy enough to make the changes related to the text tower support separate from the grad caching if everyone feels this code should be added...

**Comment by iejMac** [comment]
@rwightman The working version is on the dev-integration branch of my fork. We separated them since that one is based on your grad caching PR. 

The few experiments on cc3m that I've done with this give plausible results (i.e. nothing crazy happens) except for the examples that I showed in the LAION discord where loss explodes.

If people are interested in this I can try to remove the grad caching stuff from dev-integration, merge it into here and then it should be ready.

**Comment by rom1504** [comment]
I think that would be helpful

Not sure when exactly, but soon enough we might have some slots for experiments and it could be interesting to use that to train a multilingual clip

**Comment by BlinkDL** [comment]
I have some more ideas that might help.

Idea 1:

* Pretrained text encoder + pretrained image encoder (maybe use self-supervised training as in https://paperswithcode.com/sota/self-supervised-image-classification-on to avoid using imagenet labels)

* Freeze both encoder except their last layers, train for a while

* Unfreeze them and continue training

Idea 2 (curriculum learning for text encoder):

* Bag-of-word text encoder (sum of embedding) + usual image encoder, train for a while

* Full text encoder using the embedding + freeze image encoder, train for a while

* Unfreeze them and continue training

Idea 3 (curriculum learning for both encoders, probably overkill):

* Bag-of-word text encoder (sum of embedding) + low resolution image encoder, train for a while

* Full text encoder using the embedding + freeze image encoder, train for a while

* Freeze text encoder + full resolution image encoder, train for a while

* Unfreeze them and continue training

**Comment by rom1504** [comment]
This needs to be redone based on master.
Let's close and redo another PR or do you prefer to rebase here @iejMac ?

**Comment by iejMac** [comment]
@rom1504 @rwightman  what I currently have in this PR works on the stability cluster with the example files I temporarily added - test-roberta.json (config), hf_run.sh (training script)

still needs testing but yeah loss goes and there are no errors

**Comment by rom1504** [comment]
CI failing, some import error

**Comment by rwightman** [comment]
> CI failing, some import error

looks like PretrainedCfg needs to be defined as a dummy class for the typing arg in case where transformers isn't installed

**Comment by rom1504** [comment]
Code LGTM

Very little has been changed in the main path (without HF encoder), so this looks pretty safe

I'd be of opinion to merge this

**Comment by iejMac** [comment]
confirmed resuming works fine

**Comment by rom1504** [comment]
LGTM, resume works, inference and training also work
Very little change to main path so risk of breakage is low.

Let's merge.

We can apply fix later on if needed.
